### PR TITLE
docs(kserve-module): add upgrade service-impact test tooling and guide

### PR DESCRIPTION
**Pull request by Jooho**
Upgrade service-impact test tooling: scripts and manifests to measure inference availability during the in-tree to kserve-module operator upgrade, plus an HTML report generator and guide.

Self-contained — no dependency on the other doc PRs.

fix:  https://redhat.atlassian.net/browse/RHOAIENG-76063

## Summary by CodeRabbit

* **Documentation**
  * Added a comprehensive guide for testing upgrades from in-tree KServe to the KServe module operator.
* **New Features**
  * Added automated service-impact testing with continuous inference load generation.
  * Added snapshot and diff tooling to compare cluster state before and after upgrades.
  * Added HTML report generation with request, latency, availability, resource, and pod-change summaries.
  * Added sample test workloads and model input manifests for LLM and sklearn inference services.

### CODE REVIEW

File: kserve-module/docs/tests/upgrade/README.md [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+export VERSION_OLD=3.5.0-ea2
+export VERSION_NEW=3.5.0
+
+podman login quay.io
+```
+
+Clone the opendatahub-operator repo:
+
+```bash
+git clone https://github.com/opendatahub-io/opendatahub-operator.git $WORK/repos/opendatahub-operator
```
Comment (mwaykole) [inline kserve-module/docs/tests/upgrade/README.md:64]: Nit: trailing `/` after the closing code fence — should be just ` ``` `.

File: kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+  oc delete pod -n "$ns" curl-util --ignore-not-found 2>/dev/null
+  oc delete configmap -n "$ns" -l app=load-generator --ignore-not-found 2>/dev/null
+
+  # Delete ISVCs and runtimes
+  oc delete inferenceservice --all -n "$ns" --ignore-not-found 2>/dev/null
+  oc delete servingruntime --all -n "$ns" --ignore-not-found 2>/dev/null
+
+  # Kill log processes
+  for pidfile in "$OUTPUT_DIR"/log-pid-*; do
+    [[ -f "$pidfile" ]] || continue
```
Comment (mwaykole) [inline kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh:391]: Cleanup doesn't delete `llminferenceservice` resources. When mode is `all`, the LLMISVC deployed by `do_setup` will be left behind.

```suggestion
  oc delete inferenceservice --all -n "$ns" --ignore-not-found 2>/dev/null
  oc delete llminferenceservice --all -n "$ns" --ignore-not-found 2>/dev/null
  oc delete servingruntime --all -n "$ns" --ignore-not-found 2>/dev/null
```

File: kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+  log "Cleaning up test resources in namespace: $ns"
+
+  # Stop any running load generators and utility pods
+  oc delete pod -n "$ns" -l app=load-generator --ignore-not-found 2>/dev/null
+  oc delete pod -n "$ns" curl-util --ignore-not-found 2>/dev/null
+  oc delete configmap -n "$ns" -l app=load-generator --ignore-not-found 2>/dev/null
+
+  # Delete ISVCs and runtimes
+  oc delete inferenceservice --all -n "$ns" --ignore-not-found 2>/dev/null
+  oc delete servingruntime --all -n "$ns" --ignore-not-found 2>/dev/null
```
Comment (mwaykole) [inline kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh:387]: The ConfigMaps created in `create_load_generator_pod` (`oc create configmap "load-gen-${name}"`) have no labels, so this label-based delete won't match them. Either add `--selector` labels at creation time or delete by name pattern here:

```suggestion
  oc delete configmap -n "$ns" load-gen-sklearn-iris load-gen-llmisvc --ignore-not-found 2>/dev/null
```

File: kserve-module/docs/tests/upgrade/README.md (hunk 1 above)
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/README.md:64]: Fixed - dropped the stray `/` after the fence.

File: kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh (hunk 2 above)
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh:391]: Added `oc delete llminferenceservice --all` - it was left behind on `--mode all`.

File: kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh (hunk 3 above)
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh:387]: Switched to name-based delete (`load-gen-sklearn-iris load-gen-llmisvc`). The configmaps aren't labeled, so the `app=load-generator` selector never matched. Thanks.

File: kserve-module/docs/tests/upgrade/generate-report.sh [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                    try:
+                        d1 = datetime.fromisoformat(fail_streak_start.replace('Z','+00:00'))
+                        d2 = datetime.fromisoformat(ts.replace('Z','+00:00'))
+                        max_down_sec = max(max_down_sec, (d2 - d1).total_seconds())
+                    except: pass
+                    fail_streak_start = None
+            else:
+                buckets[ts_sec]['fail'] += 1
+                if not fail_streak_start:
+                    fail_streak_start = ts
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/generate-report.sh:80]: Internal manual test tool - report edge cases arent worth hardening here. Leaving as-is.

File: kserve-module/docs/tests/upgrade/generate-report.sh [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+before_ts = read_ts(before_name)
+after_ts = read_ts(after_name)
+
+# --- Build HTML ---
+
+total_reqs = sum(s['total'] for s in load_stats.values())
+total_fail = sum(s['fail'] for s in load_stats.values())
+max_downtime = max((s['max_downtime_s'] for s in load_stats.values()), default=0)
+overall_pass = total_fail == 0
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/generate-report.sh:205]: Same - throwaway test report, out of scope for this tool.

File: kserve-module/docs/tests/upgrade/generate-report.sh [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        <div><span class="stat-label">OK</span><span class="stat-value ok">{s['ok']:,}</span></div>
+        <div><span class="stat-label">Failed</span><span class="stat-value {'fail' if s['fail']>0 else ''}">{s['fail']:,} ({s['fail_pct']}%)</span></div>
+        <div><span class="stat-label">Max Downtime</span><span class="stat-value {'fail' if s['max_downtime_s']>0 else ''}">{s['max_downtime_s']}s</span></div>
+        <div><span class="stat-label">p50</span><span class="stat-value">{s['p50']}ms</span></div>
+        <div><span class="stat-label">p95</span><span class="stat-value">{s['p95']}ms</span></div>
+        <div><span class="stat-label">p99</span><span class="stat-value">{s['p99']}ms</span></div>
+      </div>
+      <h4>Availability Timeline</h4>
+      <div class="timeline">{svg}</div>
+      <p class="timeline-label"><span class="dot ok-dot"></span> OK <span class="dot warn-dot"></span> Partial <span class="dot fail-dot"></span> Failed &nbsp; {s['first_ts']} ~ {s['last_ts']}</p>
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/generate-report.sh:242]: Local report opened by the author with controlled timestamps. Out of scope for a test tool.

File: kserve-module/docs/tests/upgrade/load-generator.sh [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+trap 'exit 0' SIGTERM SIGINT
+
+while true; do
+  start_ns=$(date +%s%N 2>/dev/null || date +%s)
+  ts=$(date -u +"%Y-%m-%dT%H:%M:%SZ" 2>/dev/null || date -Iseconds)
+
+  http_code=$(curl -sk -o /dev/null -w '%{http_code}' \
+    --connect-timeout "$TIMEOUT" --max-time "$TIMEOUT" \
+    "$URL" 2>/dev/null) || http_code=0
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/load-generator.sh:28]: Cluster-internal health check against a self-signed test endpoint - `-k` is intentional.

File: kserve-module/docs/tests/upgrade/test-manifests/llmisvc-opt-125m-cpu.yaml [hunk 8]
```diff
@@ -0,0 +1,52 @@
+apiVersion: serving.kserve.io/v1alpha2
+kind: LLMInferenceService
+metadata:
+  annotations:
+    security.opendatahub.io/enable-auth: "false"
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/test-manifests/llmisvc-opt-125m-cpu.yaml:5]: Intentional - opt-125m is a CPU test model, auth is disabled on purpose.

File: kserve-module/docs/tests/upgrade/test-manifests/llmisvc-opt-125m-cpu.yaml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    name: facebook/opt-125m
+  replicas: 1
+  router:
+    scheduler: {}
+    route: {}
+    gateway: {}
+  template:
+    containers:
+      - name: main
+        image: public.ecr.aws/q9t5s3a7/vllm-cpu-release-repo:v0.19.0
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/test-manifests/llmisvc-opt-125m-cpu.yaml:19]: Throwaway test image; digest pinning isnt needed here.

File: kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+  local curl_opts="-s -o /dev/null -w %{http_code} --connect-timeout 5 --max-time 10"
+  [[ "$scheme" == "https" ]] && curl_opts="-sk -o /dev/null -w %{http_code} --connect-timeout 5 --max-time 10"
+
+  ensure_curl_pod "$ns"
+
+  log "Testing endpoint: $url (up to ${max_retries} retries)"
+  for i in $(seq 1 "$max_retries"); do
+    local http_code
+    http_code=$(oc exec curl-util -n "$ns" -- \
+      curl $curl_opts "$url" 2>/dev/null) || http_code=0
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh:176]: Minor style; leaving as-is for this test tool.

File: kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+  oc get namespace "$ns" &>/dev/null || oc create namespace "$ns"
+
+  deploy_sklearn_runtime "$ns"
+  deploy_sklearn_isvc "$ns"
+  test_endpoint "$ns" "sklearn-iris-predictor" "$ISVC_SVC_PORT" "$ISVC_HEALTH_PATH"
+
+  if [[ "$mode" == "all" ]]; then
+    deploy_llmisvc "$ns"
+    test_endpoint "$ns" "facebook-opt-125m-single-kserve-workload-svc" "$LLMISVC_SVC_PORT" "$LLMISVC_HEALTH_PATH" "https"
+  fi
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/test-upgrade-service-impact.sh:218]: Fixed - added validation rejecting `--mode` values other than isvc|all.

File: kserve-module/docs/tests/upgrade/upgrade-diff.sh [hunk 12]
```diff
@@ -0,0 +1,340 @@
+#!/bin/bash
+set -uo pipefail
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/upgrade-diff.sh]: Fixed - `set -euo pipefail`.

File: kserve-module/docs/tests/upgrade/upgrade-diff.sh [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+        case $1 in
+            --name) name="$2"; shift 2 ;;
+            *) echo "Unknown option: $1"; usage ;;
+        esac
+    done
+    [[ -z "$name" ]] && { echo "Error: --name is required"; usage; }
+
+    local dir="${SNAPSHOT_BASE}/${name}"
+    rm -rf "$dir"
+    mkdir -p "$dir"
```
[the remark this answers is not in the document]
Comment (Jooho) [reply kserve-module/docs/tests/upgrade/upgrade-diff.sh:158]: Local tool, `--name` is author-supplied. Out of scope.

**Review by vivekk16 — APPROVED** [review]
/lgtm
/approve
